### PR TITLE
[web]Update @dart annotation. Change shaders to allocate smaller canvas

### PR DESCRIPTION
**Pull request by ferhatb**
## Description

To align origin of shader pattern to target rect, we used to allocate large canvas(s). This PR compensates for the shader pattern origin.

Also updates engine files to 1.12.

## Related Issues

Step towards addressing:
https://github.com/flutter/flutter/issues/46833

## Tests

Covered by golden tests.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvas_pool.dart [hunk 1]
```diff
@@ -597,39 +606,117 @@ class _CanvasPool extends _SaveStackTracking {
     }
   }
 
+  /// Applys path to drawing context, preparing for fill and other operations.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvas_pool.dart:609]: nit: Applies

File: lib/web_ui/lib/src/engine/canvas_pool.dart [hunk 2]
```diff
@@ -597,39 +606,117 @@ class _CanvasPool extends _SaveStackTracking {
...
+  ///
+  /// !Don't refactor _runPath/_runPathWithOffset. Latency sensitive
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvas_pool.dart:611]: nit: move the ! to the end of the sentence. Alternatively, begin with "WARNING:"

File: lib/web_ui/lib/src/engine/html/shaders/shader.dart [hunk 3]
```diff
@@ -56,8 +56,8 @@ class GradientSweep extends EngineGradient {
     double centerX = (center.dx - shaderBounds.left) / (shaderBounds.width);
     double centerY = (center.dy - shaderBounds.top) / (shaderBounds.height);
     gl.setUniform2f(tileOffset,
-        shaderBounds.left + 2 * (shaderBounds.width * (centerX - 0.5)),
-        -shaderBounds.top - 2 * (shaderBounds.height * (centerY - 0.5)));
+        2 * (shaderBounds.width * (centerX - 0.5)),
+        2 * (shaderBounds.height * (centerY - 0.5)));
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/shaders/shader.dart:60]: nit: a trailing comma could help dartfmt do a better job

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/html/shaders/shader.dart (hunk 3 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/shaders/shader.dart:60]: 3 param function. I try to do it for lists and widget trees

File: lib/web_ui/lib/src/engine/canvas_pool.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/canvas_pool.dart]: Done.

File: lib/web_ui/lib/src/engine/canvas_pool.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/canvas_pool.dart]: Done.

**Pull request merged**